### PR TITLE
[Refactor]  Move reset_on_restart within the loop reset

### LINKED ISSUE #9562: Move reset_on_restart within the loop
### Motivation

Currently, the loop reset_on_restart the progress while reloading the states. This means setting ready, started value to the processed / completed one. 

This hidden behaviour can be mis-leading and should be exposed within the reset function of the loop.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Currently, the loop reset_on_restart the progress while reloading the states. This means setting ready, started value to the processed / completed one. 

This hidden behaviour can be mis-leading and should be exposed within the reset function of the loop.

Fixes #9562

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/progress.py [hunk 1]
```diff
@@ -229,3 +229,11 @@ def reset_on_epoch(self) -> None:
...
         self.optimizer.load_state_dict(state_dict["optimizer"])
         self.optimizer_position = state_dict["optimizer_position"]
+
+    def reset_on_restart(self):
+        self.optimizer.step.current.reset_on_restart()
+        self.optimizer.zero_grad.current.reset_on_restart()
+
+    def reset_on_batch_end(self):
+        self.optimizer.step.current.reset()
+        self.optimizer.zero_grad.current.reset()
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:239]: This is unused

```suggestion
```

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -174,6 +174,8 @@ def connect(self, epoch_loop: TrainingEpochLoop):
 
     def reset(self) -> None:
         """Resets the internal state of this loop."""
+        if self.restarting:
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:177]: Shouldn't it be `not`?

```suggestion
        if not self.restarting:
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -26,6 +26,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     * Integrate `TrainingEpochLoop.total_batch_idx` ([#8598](https://github.com/PyTorchLightning/pytorch-lightning/pull/8598))
     * Avoid optional `Tracker` attributes ([#9320](https://github.com/PyTorchLightning/pytorch-lightning/pull/9320))
     * Reset `current` progress counters when restarting an epoch loop that had already finished ([#9371](https://github.com/PyTorchLightning/pytorch-lightning/pull/9371))
+    * Move `reset_on_restart` within the loop reset function ([#9561](https://github.com/PyTorchLightning/pytorch-lightning/pull/9561))
```
Comment (carmocca) [inline CHANGELOG.md:29]:
```suggestion
    * Call `reset_on_restart` in the loop's `reset` hook instead of when loading a checkpoint ([#9561](https://github.com/PyTorchLightning/pytorch-lightning/pull/9561))
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -95,6 +95,11 @@ def reset(self) -> None:
         assert self.batch_loop.optimizer_loop is not None
         self.is_last_batch = False
 
+        if self.restarting:
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:98]: Shouldn't it be `not`?

```suggestion
        if not self.restarting:
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 5]
```diff
@@ -95,6 +95,11 @@ def reset(self) -> None:
...
+            self.batch_progress.current.reset_on_restart()
+            self.scheduler_progress.current.reset_on_restart()
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: can you put these with the `if not self.restarting below`? L106

File: pytorch_lightning/trainer/progress.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:239]: Done !

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:177]: Not it is correct. We used to reset in the loop. After we talked, you asked me to move it in the reset functions to make it explicit. Here it is :)

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:98]: This is correct.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: Done !

Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:100]: Actually, I moved it up, It makes it weird to have 2 if one after the other. At least it is cleared like this.

**Review by awaelchli — COMMENTED** [review]
the linked issue is empty and the PR is missing a description. we are missing a motivation why this needs to be changed.

**Comment by tchaton** [comment]
> the linked issue is empty and the PR is missing a description. we are missing a motivation why this needs to be changed. Also, what will 2/2 include?
> 
> If it's not ready for review, let's mark the PR as draft

Hey @awaelchli, I apologise. I started and was caught into something else. I have added a description and the PR is ready.

Best,
T.C

File: tests/loops/test_loops.py [hunk 6]
```diff
@@ -693,26 +704,26 @@ def test_fit_loop_reset(tmpdir):
...
 
         assert epoch_loop.restarting
         assert epoch_loop.batch_progress.total.ready == 2
         assert epoch_loop.batch_progress.total.completed == 1  # the checkpoint was saved on train_batch_end
-        assert epoch_loop.batch_progress.current.ready == 2
-        assert epoch_loop.batch_progress.current.completed == 2
+        assert epoch_loop.batch_progress.current.ready == 2 if has_reset else 1
+        assert epoch_loop.batch_progress.current.completed == 2 if has_reset else 1
 
     # resetting from a mid-epoch checkpoint should not change progress counters
```
Comment (awaelchli) [inline tests/loops/test_loops.py:720]: the comment here says it. the mid epoch checkpoint here is a complete one (no failure) and therefore the counters are never different. resetting should not change any values.

File: tests/loops/test_loops.py [hunk 7]
```diff
@@ -693,26 +704,26 @@ def test_fit_loop_reset(tmpdir):
...
+        assert fit_loop.epoch_progress.current.ready == (not has_reset)
         assert fit_loop.epoch_progress.current.completed == 0
 
         assert epoch_loop.restarting
         assert epoch_loop.batch_progress.total.ready == 2
         assert epoch_loop.batch_progress.total.completed == 1  # the checkpoint was saved on train_batch_end
-        assert epoch_loop.batch_progress.current.ready == 2
-        assert epoch_loop.batch_progress.current.completed == 2
+        assert epoch_loop.batch_progress.current.ready == 2 if has_reset else 1
+        assert epoch_loop.batch_progress.current.completed == 2 if has_reset else 1
```
Comment (awaelchli) [inline tests/loops/test_loops.py:718]: this should be 2 not one. we trained for 2 batches, then saved a checkpoint. there is no reason why it should be 1 here.

**Review by awaelchli — COMMENTED** [review]
I do not understand the changes you made to the test I wrote. Can you explain it?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:98]: I didn't understand your comment

File: tests/loops/test_loops.py (hunk 6 above)
Comment (tchaton) [reply tests/loops/test_loops.py:703]: Yes, it would now as the restarting logic is moved in reset

Comment (tchaton) [reply tests/loops/test_loops.py:703]: Before it would be done automatically while reloading. @carmocca argued it should be moved in reset function

Comment (awaelchli) [reply tests/loops/test_loops.py:703]: the way the test changes here just makes it not meaningful anymore. we want our tests to reflect the expected behavior and to avoid regressions. 

I'm changing it

**Review by awaelchli — APPROVED** [review]

File: tests/loops/test_loops.py (hunk 6 above)
Comment (carmocca) [reply tests/loops/test_loops.py]: @awaelchli there is value in having the checks before reset to make sure we load exactly what was saved

Comment (awaelchli) [reply tests/loops/test_loops.py]: that's already covered in other tests

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:98]: Thoughts about this @awaelchli? need to be careful here because `_num_training_batches_reached` depends on the progress value

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:98]: No it is not an elif condition. It is a double if condition.
